lib/pf/UnifiedApi/Controller/Config/Kafka.pm

```diff
@@ -0,0 +1,684 @@
+    }
+
+    return $allowed;
+}
+
+=head2 field_default
+
+Get the default value of a field
+
+=cut
+
+sub field_default {
+    my ($self, $field, $inheritedValues, $type) = @_;
+    if ($type eq 'array') {
+        return [];
+    }
+    my $default = $field->get_default_value;
+    return $default // (ref($inheritedValues) eq 'HASH' ? $inheritedValues->{$field->name} : $inheritedValues);
+}
+
+=head2 field_extra_meta
+
+Get the extra meta data for a field
+
+=cut
+
+sub field_extra_meta {
+    my ($self, $field, $meta, $parent_meta) = @_;
+    my %extra;
+    my $type = $meta->{type};
+    if ($type eq 'array') {
+        $extra{item} = $self->field_meta_array_items($field, undef, 1);
+    } elsif ($type eq 'object') {
+        $extra{properties} = $self->field_meta_object_properties($field, $meta);
+    } else {
+        if ($field->isa("HTML::FormHandler::Field::Text")) {
+            $self->field_text_meta($field, \%extra);
+        }
+
+        if ($field->isa("HTML::FormHandler::Field::Integer") || $field->isa("HTML::FormHandler::Field::IntRange")) {
+            $self->field_integer_meta($field, \%extra);
+        }
+    }
+    if ($field->has_required_when) {
+        my $required_when = $self->field_required_when($field, $meta, $parent_meta);
+        if (defined $required_when) {
+            $extra{required_when} = $required_when;
+        }
+    }
+
+    return %extra;
+}
+
+=head2 field_meta_object_properties
+
+Get the properties of a field
+
+=cut
+
+sub field_meta_object_properties {
+    my ($self, $field, $meta) = @_;
+    my %p;
+    for my $f ($field->fields) {
+        next if $field->inactive;
+        $p{$f->name} = $self->field_meta($f, $meta);
+    }
+
+    return \%p;
+}
+
+=head2 field_text_meta
+
+Update text field meta data
+
+=cut
+
+sub field_text_meta {
+    my ($self, $field, $extra) = @_;
+    my $min = $field->minlength;
+    my $max = $field->maxlength;
+    if ($min) {
+        $extra->{min_length} = $min;
+    }
+
+    if (defined $max) {
+        $extra->{max_length} = $max;
+    }
+
+    my $pattern = $field->get_tag("option_pattern");
+    if ($pattern) {
+        $extra->{pattern} = $pattern;
+    }
+
+    return ;
+}
+
+=head2 field_allowed_lookup
+
+field_allowed_lookup
+
+=cut
+
+my %FB_MODEL_2_PATH = (
+    Combination       => 'combinations',
+    Device            => 'devices',
+    DHCP6_Enterprise  => 'dhcp6_enterprises',
+    DHCP6_Fingerprint => 'dhcp6_fingerprints',
+    DHCP_Fingerprint  => 'dhcp_fingerprints',
+    DHCP_Vendor       => 'dhcp_vendors',
+    MAC_Vendor        => 'mac_vendors',
+    User_Agent        => 'user_agents',
+);
+
+sub field_allowed_lookup {
+    my ($self, $field) = @_;
+    my $allowed_lookup  = $field->get_tag("allowed_lookup") || undef;
+    if ($allowed_lookup) {
+        return $allowed_lookup;
+    }
+
+    if ($field->isa("pfappserver::Form::Field::FingerbankSelect") || $field->isa("pfappserver::Form::Field::FingerbankField")) {
```

Comment:
I am surprised to see references of fingerbank in Kafka config , is it normal ?

Reply:
This was a copy and paste from Controller::Config.
I need to refactor the shared logic into a seperate module.
This will be done in another release.